E2E: migrate Editor: Revisions to Playwright.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR migrates the Selenium suites of `wp-calypso-gutenberg-post-editor-spec.js` to Playwright.

Key changes:
- implement new objects such as `RevisionsComponent`.
- migration of test suite to Playwright.

#### Testing instructions

WPCOM (build configuration has permanent failures unrelated to this change)
- [x] mobile
- [x] desktop

Closes https://github.com/Automattic/wp-calypso/issues/55952.

worldomonation committed:
Implement basic test:
- spec file.
- changes to EditorSettingsSidebarComponent and GutenbergEditorComponent.
- new component RevisionsComponent.

worldomonation committed:
Revised:
- do not check for modal presence in EditorSettingsSidebarComponent.
- change selector used in RevisionsComponent.
- use ParagraphBlock when interacting with text.

worldomonation commented:
Incorporate #48847 into this if possible.

worldomonation commented:
Tests passed on WPCOM Playwright E2E for both mobile and desktop.

dpasque commented on test/e2e/specs/specs-wpcom/wp-editor__revisions.ts:
```diff
@@ -0,0 +1,71 @@
+		await gutenbergEditorPage.openSettings();
+		editorSettingsSidebarComponent = new EditorSettingsSidebarComponent( frame, page );
+		await editorSettingsSidebarComponent.clickTab( 'Post' );
+		await editorSettingsSidebarComponent.showRevisions();
```
Can we add a case here where select the revisions in the sidebar of the revisions modal and verify that the text changes in the modal preview before clicking load? That seems like a low-cost, low-risk, but valuable addition to this spec!

Right now, if that display pane in the modal busted, we wouldn't catch it with this current test!

worldomonation replied:
Added step to check contents of the revision diff viewer in 34af05384435710ef0a1e8e95a2b4916646084da.

dpasque commented on test/e2e/specs/specs-wpcom/wp-editor__revisions.ts:
```diff
@@ -0,0 +1,71 @@
+		await gutenbergEditorPage.visit( 'post' );
+	} );
+
+	it( 'Enter post title', async function () {
+		gutenbergEditorPage = new GutenbergEditorPage( page );
+		await gutenbergEditorPage.enterTitle( DataHelper.getRandomPhrase() );
+		await gutenbergEditorPage.saveDraft();
+	} );
+
+	it.each( [ { revision: 1 }, { revision: 2 } ] )(
+		'Create revision $revision',
+		async function ( { revision } ) {
+			const blockHandle = await gutenbergEditorPage.addBlock(
+				ParagraphBlock.blockName,
+				ParagraphBlock.blockEditorSelector
+			);
+			const paragraphBlock = new ParagraphBlock( blockHandle );
+			await paragraphBlock.enterParagraph( `Revision ${ revision }` );
+			await gutenbergEditorPage.saveDraft();
+		}
+	);
```
Total nit here: but these revision numbers totally threw me off on first read! Could we perchance make them more accurately reflect their revision number?

The first post title addition and save is technically the first revision. The two revisions added in the next case (that actually modifies the post body) are technically revisions 2 and 3!

I think where it gets confusing is that later you select revision 1, and it isn't actually the content that you labeled as "Revision 1" here

worldomonation replied:
I've double checked how revisions modal look like, and I don't see any label similar to `Revision x`:

![image](https://user-images.githubusercontent.com/6549265/150850950-72009f3d-df49-4b40-ba55-5d8d2e32fa7a.png)

Are you seeing something different?

worldomonation replied:
Ah, I think I see what you mean.

The confusion was over the creation of `Revision 1` and `Revision 2` when in fact they are meant to be `Revision 2` and `Revision 3` since the original post exists.

worldomonation replied:
I think I've fixed this in 34af05384435710ef0a1e8e95a2b4916646084da.

dpasque replied:
Oh yes, sorry, I could have been more clear! I was specifically talking about the text naming of the different revisions. The way you've restructure the test makes this a lot clearer -- I love it!

dpasque commented on test/e2e/specs/specs-wpcom/wp-editor__revisions.ts:
```diff
@@ -0,0 +1,71 @@
+/**
+ * @group gutenberg
+ */
```
Can we add the `calypso-pr` group too? This revision modal is actually added by Calypso, and is not part of Core!

worldomonation replied:
Added in 34af05384435710ef0a1e8e95a2b4916646084da.

dpasque reviewed (changes requested):
Few comments!

worldomonation committed:
Rebase onto recent trunk and incorporate major changes.

PR review comments:
- add test step to verify the contents shown in the diff viewer.
- refer to correct revision when restoring.
- add `calypso-pr` group.

dpasque commented on packages/calypso-e2e/src/lib/components/revisions-component.ts:
```diff
@@ -0,0 +1,57 @@
+	 * @param {string} text Text to check against the diff viewer.
+	 */
+	async validateTextInRevision( text: string ): Promise< void > {
+		await this.page.waitForSelector( `div.editor-revisions__dialog :text("${ text }")` );
```
Total nit, but an optional suggestion to include this among the other selectors up at the file top!

worldomonation replied:
I didn't do that because of Bart's suggestion that single-use selectors can stay within the code 😆

worldomonation replied:
Fixed in 3cc49e3203f03bef9d46e6180335a28fe224086c.

dpasque reviewed (approved):
One tiny optional nit, otherwise it looks awesome! 👍

worldomonation committed:
Move selector up
